pkg/clusterconditions/cache/cache.go

```diff
@@ -0,0 +1,169 @@
+
+	defer func() {
+		if result, ok := c.MatchResults[key]; ok {
+			result.Access = now
+		} else {
+			c.MatchResults[key] = &MatchResult{Access: now}
+		}
+	}()
+
+	for key, value := range c.MatchResults {
+		age := now.Sub(value.When)
+		aspect := "result"
+		if value.When.IsZero() {
+			age = now.Sub(value.Access)
+			aspect = "queued request"
+		}
+		if age > c.Expiration {
+			klog.V(4).Infof("pruning %q from the condition cache, as the %s is %s old", key, aspect, age)
+			delete(c.MatchResults, key)
+		}
+	}
+
+	sinceLastMatch := now.Sub(c.LastMatch)
+	if sinceLastMatch <= c.MinBetweenMatches {
+		if result, ok := c.MatchResults[key]; ok && !result.When.IsZero() {
+			return result.Match, result.Error
+		}
+		return false, fmt.Errorf("client-side throttling: only %s has elapsed since the last match call completed for this cluster condition backend; this cached cluster condition request has been queued for later execution", sinceLastMatch)
+	}
+
+	var thiefKey string
+	var thiefResult *MatchResult
+	for candidateKey, value := range c.MatchResults {
+		if !value.Access.After(value.When) {
+			continue // no requests since its last wrapped Match call
+		}
+		if thiefResult == nil ||
+			value.When.Before(thiefResult.When) || // refresh the most stale
+			(candidateKey == key && value.When == thiefResult.When) || // break ties in favor of the current request
+			value.Access.Before(thiefResult.Access) { // break ties in favor of the longest-queued
+			thiefKey = candidateKey
+			thiefResult = c.MatchResults[candidateKey]
+		}
+	}
+
+	if thiefResult != nil && now.Sub(thiefResult.When) < c.MinForCondition {
+		if result, ok := c.MatchResults[key]; ok && !result.When.IsZero() {
+			return result.Match, result.Error
+		}
+		thiefKey = key
+		thiefResult = nil
+	}
+
+	targetCondition := condition
+	if thiefKey == "" {
+		thiefKey = key
+		thiefResult = nil
+	} else if thiefKey != key {
+		if err := json.Unmarshal([]byte(thiefKey), &targetCondition); err != nil {
+			klog.V(4).Infof("%s attempted to steal this cluster-condition match call, but key is invalid JSON: %v", thiefKey, err)
+			delete(c.MatchResults, thiefKey)
+			targetCondition = condition
+			thiefKey = key
+			thiefResult = nil
+		}
+	}
+
+	if thiefKey != key {
+		var reason string
+		if thiefResult.When.IsZero() {
+			reason = "it has never been evaluated"
+		} else {
+			reason = fmt.Sprintf("its last evaluation completed %s ago", now.Sub(thiefResult.When))
+		}
+		klog.V(4).Infof("%s is stealing this cluster-condition match call for %s, because %s", thiefKey, key, reason)
+	}
+
+	match, err := c.Condition.Match(ctx, targetCondition)
+	now = time.Now()
+	c.LastMatch = now
+	if _, ok := c.MatchResults[thiefKey]; !ok {
+		c.MatchResults[thiefKey] = &MatchResult{}
+	}
+	result := c.MatchResults[thiefKey]
+	result.When = now
+	result.Match = match
+	result.Error = err
+
+	if result, ok := c.MatchResults[key]; ok && !result.When.IsZero() {
+		return result.Match, result.Error
+	}
```

Comment:
What am I missing here? I thought this would have already been covered up at line ~101.

Reply:
There are a few places we could bail and return a cached result:

1. `sinceLastMatch <= c.MinBetweenMatches`, i.e. "we are not allowed to call `Match` at all".
2. `thiefResult != nil && now.Sub(thiefResult.When) < c.MinForCondition`, i.e. "thief is our most stale, and it's still too fresh to `Match` again"
3. Down here at the bottom, after we've called `Match` for `targetCondition` (which may have been the requested condition, or may have been a thief).

Reply:
>     3. Down here at the bottom, after we've called `Match` for `targetCondition` (which may have been the requested condition, or may have been a thief).

I thought it 'd be for the above but not using `key` since I don't see it being reset. I see `thiefKey` conditionally being set to `key` so thought this would use `thiefKey` is that were the case. If not you can just explain in the call tomorrow.

Reply:
I've pushed a0b1136c -> 6dfc4186, splitting out `expireStaleMatchResults` and `calculateMostStale` helpers to make the main `Match` more compact.  Hopefully that helps avoid confusion like this^^ going forward.